# providencePipeline

A pipeline that will confirm the sequence of both the original construct as well as the transcribed product, identifying any variation from the expected sequence

## Overview

## Dependencies

* [bbmap 38.75](https://sourceforge.net/projects/bbmap/)
* [bwa 0.7.12](https://sourceforge.net/projects/bio-bwa/files/bwa-0.7.12.tar.bz2)
* [samtools 1.9](https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2)
* [bcftools 1.9](https://github.com/samtools/bcftools/releases/download/1.9/bcftools-1.9.tar.bz2)
* [vcftools 0.1.16](https://vcftools.github.io/downloads.html)
* [seqtk 1.3](https://github.com/lh3/seqtk/archive/v1.3.tar.gz)
* [blast 2.8.1](ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/2.9.0/ncbi-blast-2.9.0+-x64-linux.tar.gz)
* [rmarkdown 2.3](https://cran.r-project.org/web/packages/rmarkdown/index.html)


## Usage

### Cromwell
```
java -jar cromwell.jar run providencePipeline.wdl --inputs inputs.json
```

### Inputs

#### Required workflow parameters:
Parameter|Value|Description
---|---|---
`fastq1`|File|Read 1 fastq file, gzipped. Can be either targeted or whole transcriptome
`fastq2`|File|Read 2 fastq file, gzipped. Can be either targeted or whole transcriptome
`ref`|File|Reference to be used for alignment
`samplePrefix`|String|Prefix for output files
`sampleLibrary`|String|Library name of the transcriptome
`sampleFlowcell`|String|Flowcell name
`sampleExt`|String|External name provided for this library
`sampleConstruct`|String|Library name of the construct used


#### Optional workflow parameters:
Parameter|Value|Default|Description
---|---|---|---


#### Optional task parameters:
Parameter|Value|Default|Description
---|---|---|---
`bbMap.modules`|String|"bbmap/38.75"|Names and versions of modules needed for bbmap
`bbMap.reference`|String|"$BBMAP_ROOT/share/bbmap/resources/adapters.fa"|Reference file for adaptoer
`bbMap.trimq`|Int|25|Quality trimming
`bbMap.mem`|Int|8|Memory allocated to trimming task
`bbMap.timeout`|Int|72|Timeout in hours for this task
`bwa.modules`|String|"bwa/0.7.12 samtools/1.9"|Names and versions of modules needed for bwa-mem
`bwa.mem`|Int|12|Memory allocated for alignment task
`bwa.timeout`|Int|72|Timeout in hours for this task
`bwa.threads`|Int|8|Threads required for alignment
`variantCalling.modules`|String|"bcftools/1.9 samtools/1.9 vcftools/0.1.16 seqtk/1.3"|Names and versions of modules needed for variant calling
`variantCalling.mem`|Int|8|Memory allocated for alignment task
`variantCalling.timeout`|Int|72|Timeout in hours for this task
`blast2ReferenceSequence.modules`|String|"blast"|Names and versions of modules needed for BLAST
`blast2ReferenceSequence.mem`|Int|8|Memory allocated for alignment task
`blast2ReferenceSequence.timeout`|Int|72|Timeout in hours for this task
`qcStats.modules`|String|"bedtools samtools/1.9 picard/2.21.2"|Names and versions of modules needed for QC
`qcStats.mem`|Int|8|Memory allocated for alignment task
`qcStats.timeout`|Int|72|Timeout in hours for this task
`orfStats.modules`|String|"emboss/6.6.0"|Names and versions of modules needed for ORF generation
`orfStats.mem`|Int|8|Memory allocated for alignment task
`orfStats.timeout`|Int|72|Timeout in hours for this task
`runReport.modules`|String|"rmarkdown/0.1 pt-report-tools/1.0"|Names and versions of modules needed
`runReport.json`|String|"~{sample}.report.json"|Summary stats in JSON format
`runReport.mem`|Int|8|Memory allocated for alignment task
`runReport.timeout`|Int|72|Timeout in hours for this task


### Outputs

Output | Type | Description
---|---|---
`bam`|File|Alignment file in BAM format
`bai`|File|Alignment file index
`vcf`|File|Information on all bases in VCF format
`consensusFasta`|File|Consenus of the sample in FASTA format
`variantOnlyVcf`|File|Mutations in VCF format
`bl2seqReport`|File|Local alignment file
`alignmentStats`|File|Stats generated for the BAM file
`readDistStats`|File|Read length distribution metrics
`insertSizeStats`|File|Insert size metrics
`bbMaplog`|File|Adapter stats
`report`|File|PDF report generated by Rmarkdown
`reportScript`|File|Rmarkdown script for manual generation of the report
`orf`|File|ORF generated by EMBOSS tools


## Niassa + Cromwell

This WDL workflow is wrapped in a Niassa workflow (https://github.com/oicr-gsi/pipedev/tree/master/pipedev-niassa-cromwell-workflow) so that it can used with the Niassa metadata tracking system (https://github.com/oicr-gsi/niassa).

* Building
```
mvn clean install
```

* Testing
```
mvn clean verify \
-Djava_opts="-Xmx1g -XX:+UseG1GC -XX:+UseStringDeduplication" \
-DrunTestThreads=2 \
-DskipITs=false \
-DskipRunITs=false \
-DworkingDirectory=/path/to/tmp/ \
-DschedulingHost=niassa_oozie_host \
-DwebserviceUrl=http://niassa-url:8080 \
-DwebserviceUser=niassa_user \
-DwebservicePassword=niassa_user_password \
-Dcromwell-host=http://cromwell-url:8000
```

## Support

For support, please file an issue on the [Github project](https://github.com/oicr-gsi) or send an email to gsi@oicr.on.ca .

_Generated with generate-markdown-readme (https://github.com/oicr-gsi/gsi-wdl-tools/)_
